File: src/android/CBPlugin.java
package com.rjfun.cordova.chartboost;

import org.json.JSONObject;

import android.util.Log;
import android.view.View;

import com.chartboost.sdk.*;
import com.chartboost.sdk.Model.CBError.CBImpressionError;
import com.rjfun.cordova.ad.GenericAdPlugin;

public class CBPlugin extends GenericAdPlugin {
    private static final String LOGTAG = "Chartboost";
    
    // options
    private static final String OPT_APP_ID = "appId";
    private static final String OPT_APP_KEY = "appKey";

    private static final String TEST_BANNER_ID = "banner/not_supported";
    private static final String TEST_INTERSTITIAL_ID = "interstitial/Home Screen";
    
    private static final String TEST_APPID = "543f9b02c26ee436f622d806";
    private static final String TEST_APPKEY = "5091e071a0e429129ea7dc5b69fe005296de40ef";
    
    private String mAppId = TEST_APPID;
    private String mAppKey = TEST_APPKEY;
    
    private boolean mAppIdSet = false;

    @Override
    protected void pluginInitialize() {
    	super.pluginInitialize();
    	
    	// TODO: any init code
        Chartboost.startWithAppId(getActivity(), mAppId, mAppKey);
	}

	@Override
	protected String __getProductShortName() {
		return "Chartboost";
	}
	
	@Override
	protected String __getTestBannerId() {
		return TEST_BANNER_ID;
	}

	@Override
	protected String __getTestInterstitialId() {
		return TEST_INTERSTITIAL_ID;
	}

	@Override
	public void setOptions(JSONObject options) {
		Log.d(LOGTAG, "setOptions");
		super.setOptions(options);
		
		if(options.has(OPT_APP_ID)) this.mAppId = options.optString(OPT_APP_ID);
		if(options.has(OPT_APP_KEY)) this.mAppKey = options.optString(OPT_APP_KEY);
	}
	
	@Override
	protected View __createAdView(String adId) {
		return this.getView();
	}
	
	@Override
	protected void __loadAdView(View view) {
	}
	
	@Override
	protected int __getAdViewWidth(View view) {
		return view.getWidth();
	}

	@Override
	protected int __getAdViewHeight(View view) {
		return 0;
	}

	@Override
	protected void __pauseAdView(View view) {
	}

	@Override
	protected void __resumeAdView(View view) {
	}

	@Override
	protected void __destroyAdView(View view) {
	}
	
	protected void validateChartboost() {
		if(! mAppIdSet) {
			Log.d(LOGTAG, "startWithAppId: " + mAppId + ", " + mAppKey);
			Chartboost.startWithAppId(getActivity(), mAppId, mAppKey);
			Chartboost.setDelegate(new CBAdEventDelegate());
			Chartboost.onCreate(getActivity());
			Chartboost.onStart(getActivity());
			mAppIdSet = true;
		}
	}
	
	@Override
	protected Object __createInterstitial(String adId) {
		Log.d(LOGTAG, "__createInterstitial");
		validateChartboost();
		
		String[] items = adId.split("/");
		if(items.length >= 2) {
			return adId;
		}
        return null;
	}
	
	@Override
	protected void __loadInterstitial(Object interstitial) {
		Log.d(LOGTAG, "__loadInterstitial");
		if(interstitial == null) return;
		
		if(interstitial instanceof String) {
			String[] items = ((String) interstitial).split("/");
			if(items.length >= 2) {
				String adType = items[0];
				String adLocation = items[1];
				
				if("video".compareTo(adType) == 0) {
					Chartboost.cacheRewardedVideo(adLocation);
					
				} else if("moreapps".compareTo(adType) == 0) {
					Chartboost.cacheMoreApps(adLocation);
					
				} else {
					Chartboost.cacheInterstitial(adLocation);
				}
			}
		}
	}
	
	@Override
	protected void __showInterstitial(Object interstitial) {
		Log.d(LOGTAG, "__showInterstitial");
		if(interstitial == null) return;
		
		if(interstitial instanceof String) {
			String[] items = ((String) interstitial).split("/");
			if(items.length >= 2) {
				String adType = items[0];
				String adLocation = items[1];
				
				if("video".compareTo(adType) == 0) {
					if(Chartboost.hasRewardedVideo(adLocation)) Chartboost.showRewardedVideo(adLocation);
					
				} else if("moreapps".compareTo(adType) == 0) {
					if(Chartboost.hasMoreApps(adLocation)) Chartboost.showMoreApps(adLocation);
					
				} else {
					if(Chartboost.hasInterstitial(adLocation)) Chartboost.showInterstitial(adLocation);
				}
			}
		}
	}

	@Override
	protected void __destroyInterstitial(Object interstitial) {
		Log.d(LOGTAG, "__destroyInterstitial");
		if(interstitial == null) return;

		// no need do anything
	}

    @Override
    public void onPause(boolean multitasking) {
    	Chartboost.onPause(getActivity());
        super.onPause(multitasking);
    }
    
    @Override
    public void onResume(boolean multitasking) {
    	super.onResume(multitasking);
    	Chartboost.onResume(getActivity());
    }

    @Override 
    public void onStart() {
    	super.onStart();
        Chartboost.onStart(getActivity());
    }
    
    @Override
    public void onDestroy() {
    	Chartboost.onDestroy(getActivity());
        super.onDestroy();
    }
    
    /**
     * document.addEventListener('onAdLoaded', function(data));
     * document.addEventListener('onAdFailLoad', function(data));
     * document.addEventListener('onAdPresent', function(data));
     * document.addEventListener('onAdDismiss', function(data));
     * document.addEventListener('onAdLeaveApp', function(data));
     */
    private class CBAdEventDelegate extends ChartboostDelegate {
    	
    	// Interstitial Delegate Methods /////////////////////////////
    	
    	// Called after an interstitial has been displayed on the screen.
        @Override
    	public void didDisplayInterstitial(String location) {
        	fireAdEvent(EVENT_AD_PRESENT, ADTYPE_INTERSTITIAL);
        }

    	// Called after an interstitial has been loaded from the Chartboost API
    	// servers and cached locally.
    	public void didCacheInterstitial(String location) {
        	fireAdEvent(EVENT_AD_LOADED, ADTYPE_INTERSTITIAL);
        	
            if(autoShowInterstitial) {
            	showInterstitial();
            }
        }

    	// Called after an interstitial has attempted to load from the Chartboost API
    	// servers but failed.
    	public void didFailToLoadInterstitial(String location, CBImpressionError error) {
    		fireAdErrorEvent(EVENT_AD_FAILLOAD, error.ordinal(), getErrorReason(error), ADTYPE_INTERSTITIAL);
        }

    	// Called after an interstitial has been dismissed.
    	public void didDismissInterstitial(String location) {
        	fireAdEvent(EVENT_AD_DISMISS, ADTYPE_INTERSTITIAL);
        }

    	// Called after an interstitial has been closed.
    	public void didCloseInterstitial(String location){
        	
        }

    	// Called after an interstitial has been clicked.
    	public void didClickInterstitial(String location) {
    		fireAdEvent(EVENT_AD_LEAVEAPP, ADTYPE_INTERSTITIAL);
        }
    	
    	// Video Delegate Methods /////////////////////////////
    	
    	// Called after a rewarded video has been loaded from the Chartboost API
    	// servers and cached locally.
    	public void didCacheRewardedVideo(String location) {
        	fireAdEvent(EVENT_AD_LOADED, "video");
        	
            if(autoShowInterstitial) {
            	showInterstitial();
            }
        }

    	// Called after a rewarded video has attempted to load from the Chartboost API
    	// servers but failed.
    	public void didFailToLoadRewardedVideo(String location, CBImpressionError error) {
    		fireAdErrorEvent(EVENT_AD_FAILLOAD, error.ordinal(), getErrorReason(error), "video");
        }

    	// Called after a rewarded video has been dismissed.
    	public void didDismissRewardedVideo(String location) {
        	fireAdEvent(EVENT_AD_DISMISS, "video");
        }

    	// Called after a rewarded video has been closed.
    	public void didCloseRewardedVideo(String location) {
    		
    	}

    	// Called after a rewarded video has been clicked.
    	public void didClickRewardedVideo(String location) {
    		fireAdEvent(EVENT_AD_LEAVEAPP, "video");
        }

    	// Called after a rewarded video has been viewed completely and user is eligible for reward.
    	public void didCompleteRewardedVideo(String location, int reward) {
    		//fireAdEvent(EVENT_AD_LEAVEAPP, ADTYPE_INTERSTITIAL);
        }

    	// MoreApps Delegate Methods /////////////////////////////
    	
    	// Called after a MoreApps page has been loaded from the Chartboost API
		// servers and cached locally.
		public void didCacheMoreApps(String location) {
        	fireAdEvent(EVENT_AD_LOADED, "moreapps");
            if(autoShowInterstitial) {
            	showInterstitial();
            }
        }
		
		// Called after a MoreApps page has been dismissed.
		public void didDismissMoreApps(String location) {
        	fireAdEvent(EVENT_AD_DISMISS, "moreapps");
        }
		
		// Called after a MoreApps page has been closed.
		public void didCloseMoreApps(String location) {
			
		}
		
		// Called after a MoreApps page has been clicked.
		public void didClickMoreApps(String location) {
    		fireAdEvent(EVENT_AD_LEAVEAPP, "moreapps");
        }
		
		// Called after a MoreApps page attempted to load from the Chartboost API
		// servers but failed.
		public void didFailToLoadMoreApps(String location, CBImpressionError error) {
    		fireAdErrorEvent(EVENT_AD_FAILLOAD, error.ordinal(), getErrorReason(error), "moreapps");
        }
    }
    
   /** Gets a string error reason from an error code. */
   public String getErrorReason(CBImpressionError error)
   {
       switch(error) {
           case INTERNAL:
               return "Unknown internal error";
           case INTERNET_UNAVAILABLE:
               return "Network is currently unavailable";
           case TOO_MANY_CONNECTIONS:
               return "Too many requests are pending for that location";
           case WRONG_ORIENTATION:
               return "Interstitial loaded with wrong orientation";
           case FIRST_SESSION_INTERSTITIALS_DISABLED:
               return "Interstitial disabled, first session";
           case NETWORK_FAILURE:
               return "Network request failed";
           case NO_AD_FOUND:
               return "No ad received";
           case SESSION_NOT_STARTED:
               return "Session not started";
           case IMPRESSION_ALREADY_VISIBLE: 
        	   return "Impression already visible";
           case NO_HOST_ACTIVITY:
        	   return "No host activity";
           case USER_CANCELLATION:
               return "User manually cancelled the impression";
           case INVALID_LOCATION:
               return "No location detected";
           case VIDEO_UNAVAILABLE:
               return "No location detected";
           case VIDEO_ID_MISSING:
               return "No location detected";
           case ERROR_PLAYING_VIDEO:
               return "No location detected";
           case INVALID_RESPONSE:
               return "No location detected";
           case ERROR_CREATING_VIEW:
               return "No location detected";
           case ERROR_DISPLAYING_VIEW:
               return "No location detected";
           case ASSETS_DOWNLOAD_FAILURE:
               return "Prefetching is not finished";
       }
       return "Unknown error";
   }

}

File: src/ios/CBPlugin.m
//
//  CDVAdMobPlugin.m
//  TestAdMobCombo
//
//  Created by Xie Liming on 14-10-20.
//
//

#import <CommonCrypto/CommonDigest.h>
#import <AdSupport/AdSupport.h>

#import <Chartboost/Chartboost.h>
#import <Chartboost/CBNewsfeed.h>

#import "CBPlugin.h"

#define TEST_BANNER_ID           @"banner/not_supported"
#define TEST_INTERSTITIALID      @"interstitial/Home Screen"

#define TEST_APPID              @"543f9ac9c26ee436e7133794"
#define TEST_APPKEY             @"6d1c2e4a5d1225825a7472aa52d0f45e4757de39"

#define BANNER_AD_WIDTH         320
#define BANNER_AD_HEIGHT        50

@interface CBPlugin()<ChartboostDelegate>

@property (nonatomic, retain) NSString* mAppId;
@property (nonatomic, retain) NSString* mAppKey;

@property (assign) bool mAppIdSet;

- (void) validateChartboost;

@end

@implementation CBPlugin

- (void)pluginInitialize
{
    [super pluginInitialize];
    
    self.adWidth = BANNER_AD_WIDTH;
    self.adHeight = BANNER_AD_HEIGHT;
    
    self.mAppId = TEST_APPID;
    self.mAppKey = TEST_APPKEY;
    
    self.testTraffic = FALSE;
    
    self.mAppIdSet = FALSE;
}

- (NSString*) __getProductShortName { return @"Chartboost"; }

- (NSString*) __getTestBannerId {
    return TEST_BANNER_ID;
}
- (NSString*) __getTestInterstitialId {
    return TEST_INTERSTITIALID;
}

- (void) validateChartboost
{
    if(! self.mAppIdSet) {
        [Chartboost startWithAppId:self.mAppId
                      appSignature:self.mAppKey
                          delegate:self];
        
        self.mAppIdSet = TRUE;
    }
}

- (void) parseOptions:(NSDictionary *)options
{
    NSLog(@"parseOptions");
    
    [super parseOptions:options];
    
    if(self.isTesting) {
    }
    
    NSString* str = [options objectForKey:OPT_APP_ID];
    if(str) self.mAppId = str;
    
    str = [options objectForKey:OPT_APP_KEY];
    if(str) self.mAppKey = str;
}

- (UIView*) __createAdView:(NSString*)adId
{
    return [self getView];
}

- (void) __showBanner:(int) position atX:(int)x atY:(int)y
{
}

- (int) __getAdViewWidth:(UIView*)view {
    return view.frame.size.width;
}

- (int) __getAdViewHeight:(UIView*)view {
    return view.frame.size.height;
}

- (void) __loadAdView:(UIView*)view
{
}

- (void) __pauseAdView:(UIView*)view {
}

- (void) __resumeAdView:(UIView*)view {
}

- (void) __destroyAdView:(UIView*)view {
    if(! view) return;
    
}

- (NSObject*) __createInterstitial:(NSString*)adId {
    NSLog(@"__createInterstitial");
    
    [self validateChartboost];
    
    NSArray* fields = [adId componentsSeparatedByString:@"/"];
    if([fields count] >= 2) {
        //NSString* adType = [fields objectAtIndex:0];
        //NSString* adLocation = [fields objectAtIndex:1];
        // we use the ad type and ad location as interstitial id
        return adId;
    }
    
    return nil;
}

- (void) __loadInterstitial:(NSObject*)interstitial {
    NSLog(@"__loadInterstitial");
    
    if(! interstitial) return;
    NSString* adId = (NSString*) interstitial;
    
    NSArray* fields = [adId componentsSeparatedByString:@"/"];
    if([fields count] >= 2) {
        NSString* adType = [fields objectAtIndex:0];
        NSString* adLocation = [fields objectAtIndex:1];
        
        if (adType!=nil && [adType compare:@"video"] == NSOrderedSame) {
            [Chartboost cacheRewardedVideo:adLocation];
            
        } else if (adType!=nil && [adType compare:@"moreapps"] == NSOrderedSame) {
            [Chartboost cacheMoreApps:adLocation];
            
        } else {
            [Chartboost cacheInterstitial:adLocation];
            
        }
    }
}

- (void) __showInterstitial:(NSObject*)interstitial {
    NSLog(@"__showInterstitial");
    
    if(! interstitial) return;
    NSString* adId = (NSString*) interstitial;
    
    NSArray* fields = [adId componentsSeparatedByString:@"/"];
    if([fields count] >= 2) {
        NSString* adType = [fields objectAtIndex:0];
        NSString* adLocation = [fields objectAtIndex:1];
        
        if (adType!=nil && [adType compare:@"video"] == NSOrderedSame) {
            if([Chartboost hasRewardedVideo:adLocation]) {
                [Chartboost showRewardedVideo:adLocation];
            }
            
        } else if (adType!=nil && [adType compare:@"moreapps"] == NSOrderedSame) {
            if([Chartboost hasMoreApps:adLocation]) {
                [Chartboost showMoreApps:adLocation];
            }
            
        } else {
            if([Chartboost hasInterstitial:adLocation]) {
                [Chartboost showInterstitial:adLocation];
            }
        }
    }
}

- (void) __destroyInterstitial:(NSObject*)interstitial {
    if(! interstitial) return;
    // NSString* adId = (NSString*) interstitial;
    // do nothing
}

#pragma mark GADBannerViewDelegate implementation

/**
 * document.addEventListener('onAdLoaded', function(data));
 * document.addEventListener('onAdFailLoad', function(data));
 * document.addEventListener('onAdPresent', function(data));
 * document.addEventListener('onAdDismiss', function(data));
 * document.addEventListener('onAdLeaveApp', function(data));
 */

- (NSString*) getErrorReason:(CBLoadError)error
{
    switch(error) {
        case CBLoadErrorInternal:
            return @"Unknown internal error";
        case CBLoadErrorInternetUnavailable:
            return @"Network is currently unavailable";
        case CBLoadErrorTooManyConnections:
            return @"Too many requests are pending for that location";
        case CBLoadErrorWrongOrientation:
            return @"Interstitial loaded with wrong orientation";
        case CBLoadErrorFirstSessionInterstitialsDisabled:
            return @"Interstitial disabled, first session";
        case CBLoadErrorNetworkFailure:
            return @"Network request failed";
        case CBLoadErrorNoAdFound:
            return @"No ad received";
        case CBLoadErrorSessionNotStarted:
            return @"Session not started";
        case CBLoadErrorUserCancellation:
            return @"User manually cancelled the impression";
        case CBLoadErrorNoLocationFound:
            return @"No location detected";
        case CBLoadErrorPrefetchingIncomplete:
            return @"Video Prefetching is not finished";
    }
    return @"Unknown error";
}

- (void)didFailToLoadInterstitial:(CBLocation)location withError:(CBLoadError)error
{
    NSLog(@"%d - %@", (int)error, [self getErrorReason:error]);

    [self fireAdErrorEvent:EVENT_AD_FAILLOAD withCode:(int)error
                   withMsg:[self getErrorReason:error]
                  withType:ADTYPE_INTERSTITIAL];
}

- (void)didCacheInterstitial:(CBLocation)location
{
    if (self.interstitial && self.autoShowInterstitial) {
        dispatch_async(dispatch_get_main_queue(), ^{
            [self __showInterstitial:[NSString stringWithFormat:@"interstitial/%@", location]];
        });
    }
    [self fireAdEvent:EVENT_AD_LOADED withType:ADTYPE_INTERSTITIAL];
}

- (void)didDisplayInterstitial:(CBLocation)location
{
    [self fireAdEvent:EVENT_AD_PRESENT withType:ADTYPE_INTERSTITIAL];
}

- (void)didDismissInterstitial:(CBLocation)location
{
    [self fireAdEvent:EVENT_AD_DISMISS withType:ADTYPE_INTERSTITIAL];
}

- (void)didCloseInterstitial:(CBLocation)location
{
    //[self fireAdEvent:EVENT_AD_DISMISS withType:ADTYPE_INTERSTITIAL];
}

- (void)didClickInterstitial:(CBLocation)location
{
    [self fireAdEvent:EVENT_AD_LEAVEAPP withType:ADTYPE_INTERSTITIAL];
}

- (void)didFailToLoadRewardedVideo:(CBLocation)location withError:(CBLoadError)error
{
    NSLog(@"%d - %@", (int)error, [self getErrorReason:error]);
    
    [self fireAdErrorEvent:EVENT_AD_FAILLOAD withCode:(int)error
                   withMsg:[self getErrorReason:error]
                  withType:@"video"];
}

- (void)didCacheRewardedVideo:(CBLocation)location
{
    if (self.interstitial && self.autoShowInterstitial) {
        dispatch_async(dispatch_get_main_queue(), ^{
            [self __showInterstitial:[NSString stringWithFormat:@"video/%@", location]];
        });
    }
    [self fireAdEvent:EVENT_AD_LOADED withType:@"video"];
}

- (void)didDisplayRewardedVideo:(CBLocation)location
{
    [self fireAdEvent:EVENT_AD_PRESENT withType:@"video"];
}

- (void)didDismissRewardedVideo:(CBLocation)location
{
    [self fireAdEvent:EVENT_AD_DISMISS withType:@"video"];
}

- (void)didCloseRewardedVideo:(CBLocation)location
{
    //[self fireAdEvent:EVENT_AD_DISMISS withType:@"video"];
}

- (void)didClickRewardedVideo:(CBLocation)location
{
    [self fireAdEvent:EVENT_AD_LEAVEAPP withType:@"video"];
}

// ---------------------------------------------------------------

- (void)didFailToLoadMoreApps:(CBLocation)location withError:(CBLoadError)error
{
    NSLog(@"%d - %@", (int)error, [self getErrorReason:error]);
    
    [self fireAdErrorEvent:EVENT_AD_FAILLOAD withCode:(int)error
                   withMsg:[self getErrorReason:error]
                  withType:@"moreapps"];
}

- (void)didCacheMoreApps:(CBLocation)location
{
    if (self.interstitial && self.autoShowInterstitial) {
        dispatch_async(dispatch_get_main_queue(), ^{
            [self __showInterstitial:[NSString stringWithFormat:@"moreapps/%@", location]];
        });
    }
    [self fireAdEvent:EVENT_AD_LOADED withType:@"moreapps"];
}

- (void)didDisplayMoreApps:(CBLocation)location
{
    [self fireAdEvent:EVENT_AD_PRESENT withType:@"moreapps"];
}

- (void)didDismissMoreApps:(CBLocation)location
{
    [self fireAdEvent:EVENT_AD_DISMISS withType:@"moreapps"];
}

- (void)didCloseMoreApps:(CBLocation)location
{
    //[self fireAdEvent:EVENT_AD_DISMISS withType:@"moreapps"];
}

- (void)didClickMoreApps:(CBLocation)location
{
    [self fireAdEvent:EVENT_AD_LEAVEAPP withType:@"moreapps"];
}

@end
